metadocs: add poc to single-sourcing/partials (#1230)
redpanda-data/documentation: README.md

In Redpanda documentation, save partials in a `shared` folder, at the root of the directory that shares the same context. If the partial doc can be used in more contexts, go up on level in the file tree. For example, the `docs/deploy/kubernetes/shared` folder contains partials with k8s content that's imported by docs in `docs/deploy/kubernetes`; general partials are stored at `docs/shared` and used throughout the project.

Partials can be imported into other partials.

Example:

File `docs/deploy/shared/_markdown-partial-example.mdx`:

```javascript
<span>Hello {props.name}</span>
This is text some content from `_markdown-partial-example.mdx`.
```

The file that imports it must follow this syntax:

File `docs/deploy/use-example.mdx`:

```javascript
import PartialExample from './shared/_markdown-partial-example.mdx';
<PartialExample name="Redpanda Docs" />
```

The HTML output of the file `docs/deploy/use-example.mdx` will be:

```
Hello Redpanda Docs
This is text some content from _markdown-partial-example.md.
```

### Limitations :warning:

> - **Headings**. Currently Docusaurus has a limitation in which the table of contents (the right navigation pane) doesn't import headings in partials. See its [Docusaurus open issue](https://github.com/facebook/docusaurus/issues/3915). There's no know workaround.

Reviewer: ```suggestion
- **Headings**. Currently Docusaurus has a limitation in which the table of contents (the right navigation pane) doesn't import headings in partials. See its [Docusaurus open issue](https://github.com/facebook/docusaurus/issues/3915). There is no known workaround.
```

> - **Relative links**. Relative links still follows the file tree where they were created. The partials feature doesn't change the link destination. For example, if your partial is saved in `docs/folderA/shared/_partial` and you want import it from `docs/folderB/folderC/my-file`, the relative link won't work, because the relative location is physically different. 

Reviewer: ```suggestion
- **Relative links**. Relative links in partials are not modified to account for the paths of the files that import them. For example, if your partial is saved in `docs/folderA/shared/_partial` and you want import it from `docs/folderB/folderC/my-file`, the relative link won't work, because the relative location is different. 
```

Author: I love this change !

>     :bulb: Workaround: Use absolute links or use links as variables.

Reviewer: I don't think we have a working example yet of a link as a variable, so let's remove that statement for now.

```suggestion
    :bulb: Workaround: Use absolute links.
```

Reviewer 2: I don't think we should use absolute links. It will break when the content is versioned.

Author: This was before your proposed linking scheme. This section should be replaced.

## Style Guide

Before you add or edit content, consult the Redpanda Style Guide (available on the [wiki](https://vectorizedio.atlassian.net/wiki/spaces/DOC/pages/92635182/Redpanda+Style+Guide) or in [Google Docs](https://drive.google.com/drive/folders/1dZqaWhAqe5-jHcOd0y6WRflMUxEPNK4Q?ths=true)) for product documentation guidelines.  

> ## Single-sourcing/ partials
> There are some instances where you want to write one document and display in two (or more) pages. For that use case, you want your content to be single-sourced. Examples: common steps, introductory section that needs to be repeated.
> 
> ### How
> Docusaurus has a feature where you can import markdown pages into one another. They call those files **partials**. 
> 
> While it's technically possible to import any page into any other page, we have some convetions.
> 
> - Any imported page use `_` as the filename prefix. 
> - Use a folder called `shared` to put your partials. You have to decide in what level this folder should be created. If you think that this partial page could be used by other contexts, create the folder in the level that makes sense. For example, the product version makes sense to the whole project, so we could create it at `docs/shared`. Some niche kubernetes docs only makes sense in that context, then `docs/deploy/kubernetes/shared`. 
> 
> Examples on how to use it:
> 
> File `docs/shared/_markdown-partial-example.mdx`:
> 
> ```javascript
> <span>Hello {props.name}</span>
> 
> This is text some content from `_markdown-partial-example.mdx`.
> ```
> 
> The file that imports it must follow this syntax: 
> 
> File `docs/deploy/use-example.mdx`:
> 
> ```javascript
> import PartialExample from './_markdown-partial-example.mdx';
> 
> <PartialExample name="Redpanda Docs" />
> ```
> 
> The output of the file `docs/deploy/use-example.mdx` will be:
> 
> ```
> Hello Redpanda Docs
> This is text some content from _markdown-partial-example.md.
> ```
> 
> More on that, on [Docusaurus documentation on partials](https://docusaurus.io/docs/markdown-features/react#importing-markdown)
> 
> ### Limitations
> Currently Docusaurus has a huge limitation in which the table of contents (the right side nav), does not contain the imported Markdown headings. [They have an open issue tracking it](https://github.com/facebook/docusaurus/issues/3915). There's no know workaround.
> 
> Relative links will only work where they were created. For example, if your partial file lives under `docs/folderA/shared/_partial` and you want to use in `docs/folderB/folderC/my-file` the relative link won't work, because the relative location is different, you have essentially another level. Workaround: Use fixed links, use links as variables.
> 
> ### Examples
> 
> - [docs/get-started/shared/_rpk-version.mdx](https://github.com/redpanda-data/documentation/blob/dev/docs/get-started/shared/_rpk-version.mdx)
> - [docs/manage/kubernetes/shared/_values-yaml.mdx](https://github.com/redpanda-data/documentation/blob/dev/docs/manage/kubernetes/shared/_values-yaml.mdx)

Reviewer: Edited and rewritten for conciseness.

```suggestion
## Best practices

### Single-sourcing with partials

Practice the DRY (Don't Repeat Yourself) principle by single-sourcing repeated content. Common examples of single-sourced content include prerequisites, contact info, and foundational steps of how-to guides.  

Docusaurus supports single-sourcing with partials: markdown files that are importable in other files. The markdown content in a partial is rendered in a file that imports it. See [Docusaurus](https://docusaurus.io/docs/markdown-features/react#importing-markdown) for an example partial and an example import of a partial.

Partial filenames must be prefixed with an underscore (`_my_partial.mdx`). Docusaurus doesn't create doc pages for files prefixed with an underscore.

In Redpanda documentation, save partials in a `shared` folder, with `shared` at the root of the directory tree containing the files that import the partials. For example, in Redpanda documentation the `docs/deploy/kubernetes/shared` folder contains partials with k8s content that's imported by docs in `docs/deploy/kubernetes`.

> :warning: **Don't include links or headings in partials**. Currently Docusaurus has a limitation in which the table of contents in the right navigation pane doesn't display headings in partials. See its [Docusaurus open issue](https://github.com/facebook/docusaurus/issues/3915).

> :bulb: **Use absolute links in partials**. Relative links in partials will only work where they were created. For example, if your partial is saved in `docs/folderA/shared/_partial` and you want import it from `docs/folderB/folderC/my-file`, the relative link won't work, because the relative location is different.

Examples of partials in Redpanda documentation:

- [docs/get-started/shared/_rpk-version.mdx](https://github.com/redpanda-data/documentation/blob/dev/docs/get-started/shared/_rpk-version.mdx)
- [docs/manage/kubernetes/shared/_values-yaml.mdx](https://github.com/redpanda-data/documentation/blob/dev/docs/manage/kubernetes/shared/_values-yaml.mdx)
```

Author: > Docusaurus supports single-sourcing with partials: markdown files that are importable in other files. The markdown content in a partial is rendered in a file that imports it. See [Docusaurus](https://docusaurus.io/docs/markdown-features/react#importing-markdown) for an example partial and an example import of a partial.

It doesn't have to be a partial to be imported. Any file can be imported. We use partials purely out of convention.

Author: Also, I don't want to remove the example of our own docs. Even though docusaurus has the same example, we want to have all the context in this doc.
